### PR TITLE
Deterministic ordering of competing messages

### PR DESCRIPTION
**Pull request by abdulmth**
# Description of change
Ordering message now is implemented according to the specs, namely: 

If two or more root documents, integration documents or diff documents compete in the chain. Eg. 2 integration message have the same `previous_message_id` they will be sorted according to the milestone that references them. If both messages have the same milestone, they will be alphabetically ordered according based on their `MessageId`. 

**Note: Milestone ordering only happens if two or more message compete in the chain!**

## Links to any relevant issues
fixes issue #480

## Type of change

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [ ] Enhancement (a non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Fix

### CODE REVIEW

**Review by cycraig — APPROVED** [review]
Looks fine to me after the changes.

Passing around the `Client` is not ideal but I don't think we can get around that without fetching all milestones up-front, which is a lot more expensive in general.

Minor comments about errors.

**Review by olivereanderson — APPROVED** [review]
Looks mostly good to me. I agree with what @cycraig said regarding `MilestoneError` though.

File: identity-iota/src/chain/milestone.rs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        .client
+        .get_message()
+        .metadata(message.message_id())
+        .await
+        .map(|metadata| (metadata.referenced_by_milestone_index, message))
+    })
+    .collect::<FuturesUnordered<_>>()
+    .try_collect()
+    .await
+    .map_err(|_| MilestoneError)?; // TODO: return actual error / error message too?
```
[the remark this answers is not in the document]
Comment (cycraig) [reply identity-iota/src/chain/milestone.rs]: Removed `MilestoneError`, now it just bubbles up `ClientError` all the way.

**Review by olivereanderson — APPROVED** [review]

File: identity-iota/src/chain/integration_chain.rs [hunk 2]
```diff
@@ -33,40 +35,53 @@ pub struct IntegrationChain {
...
-      'inner: while let Some(document) = list.pop() {
-        if this.try_push(document).is_ok() {
-          break 'inner;
-        }
+    let root_document: IotaDocument = {
+      let valid_root_documents: Vec<IotaDocument> = index
+        .remove(&MessageId::null())
+        .ok_or(Error::ChainError {
+          error: "DID not found or pruned",
+        })?
```
[the remark this answers is not in the document]
Comment (cycraig) [reply identity-iota/src/chain/integration_chain.rs]: Added `DIDNotFound` error variant,

**Pull request merged**